File: CSharp/Ops/McUdpSendDataHandler.cs
///////////////////////////////////////////////////////////
//  McUdpSendDataHandler.cs
//  Implementation of the Class McUdpSendDataHandler
//  Created on:      06-jan-2013
//  Author:
///////////////////////////////////////////////////////////

using System.Collections.Generic;
using System.Runtime.CompilerServices;  // Needed for the "MethodImpl" synchronization attribute

namespace Ops 
{
	public class McUdpSendDataHandler : SendDataHandler 
    {
        private class Entry_T
        {
            public bool staticRoute = false;
            //                port
            public Dictionary<string, IpPortPair> portMap = new Dictionary<string, IpPortPair>();
            public Entry_T(bool staticRoute) { this.staticRoute = staticRoute; }
        }

        //                 Topic
        private Dictionary<string, Entry_T > topicSinkMap = new Dictionary<string,Entry_T>();
        

        public McUdpSendDataHandler(int outSocketBufferSize, string localInterface) 
        {
            sender = new UdpSender(0, localInterface, outSocketBufferSize);  
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public override bool SendData(byte[] bytes, int size, Topic t)
        {
            bool result = true;

            if (this.topicSinkMap.ContainsKey(t.GetName()))
            {
                Entry_T dict = this.topicSinkMap[t.GetName()];
                List<string> sinksToDelete = new List<string>();

                // Loop over all sinks and send data, remove items that isn't "alive".
                foreach (KeyValuePair<string, IpPortPair> kvp in dict.portMap)
                {
                    // Check if this sink is alive
                    if (kvp.Value.IsAlive())
                    {
                        result &= SendData(bytes, size, InetAddress.GetByName(kvp.Value.Ip), kvp.Value.Port);
                    }
                    else //Remove it.
                    {
                        sinksToDelete.Add(kvp.Key);
                    }
                }
                foreach (string key in sinksToDelete)
                {
                    dict.portMap.Remove(key);
                }
            }
            return result;
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void AddSink(string topic, string ip, int port, bool staticRoute)
        {
            IpPortPair ipp = new IpPortPair(ip, port, staticRoute);

            if (this.topicSinkMap.ContainsKey(topic))
            {
                Entry_T dict = this.topicSinkMap[topic];

                //If created as static route, we only add sinks that are static
                if ((!dict.staticRoute) || (dict.staticRoute && staticRoute))
                {
                    //Check if sink already exist
                    if (dict.portMap.ContainsKey(ipp.Key))
                    {
                        dict.portMap[ipp.Key].FeedWatchdog(staticRoute);
                    }
                    else
                    {
                        dict.portMap.Add(ipp.Key, ipp);
                    }
                }
            }
            else
            {
                Entry_T dict = new Entry_T(staticRoute);
                dict.portMap.Add(ipp.Key, ipp);
                this.topicSinkMap.Add(topic, dict);
            }
        }

        private class IpPortPair
        {
            public string Ip {get; private set; }
            public int Port {get; private set; }
            public string Key { get; private set; }
            private bool alwaysAlive = false;

            private System.DateTime lastTimeAlive;
            private System.TimeSpan ALIVE_TIMEOUT = System.TimeSpan.FromMilliseconds(3000.0);

            public IpPortPair(string ip, int port, bool alwaysAlive)
            {
                this.Ip = ip;
                this.Port = port;
                this.Key = Ip + ":" + Port.ToString();
                this.alwaysAlive = alwaysAlive;
                this.lastTimeAlive = System.DateTime.Now;
            }

            public bool IsAlive()
            {
                return this.alwaysAlive || ((System.DateTime.Now - this.lastTimeAlive) < ALIVE_TIMEOUT);
            }

            public void FeedWatchdog(bool alwaysAlive)
            {
                this.alwaysAlive |= alwaysAlive;
                this.lastTimeAlive = System.DateTime.Now;
            }
        }
	}

}

File: Java/OPSJLib/src/ops/McUdpSendDataHandler.java
/**
*
* Copyright (C) 2006-2009 Anton Gravestam.
* Copyright (C) 2020 Lennart Andersson.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/

package ops;

import java.io.IOException;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.util.Vector;
import java.util.HashMap;

public class McUdpSendDataHandler extends SendDataHandlerBase
{
    private class Entry_T
    {
        public boolean staticRoute = false;
        public HashMap<String, IpPortPair> portMap = new HashMap<String, IpPortPair>();
        public Entry_T(boolean staticRoute) { this.staticRoute = staticRoute; }
    }

    //              Topic
    private HashMap<String, Entry_T> topicSinkMap = new HashMap<String,Entry_T>();

    public McUdpSendDataHandler(int outSocketBufferSize, String localInterface) throws IOException
    {
        sender = new UDPSender(0, localInterface, outSocketBufferSize);
    }

    public synchronized boolean sendData(byte[] bytes, int size, Topic t)
    {
        boolean result = true;

        if (this.topicSinkMap.containsKey(t.getName()))
        {
            Entry_T dict = this.topicSinkMap.get(t.getName());
            Vector<String> sinksToDelete = new Vector<String>();

            // Loop over all sinks and send data, remove items that isn't "alive".
            for (IpPortPair ipp : dict.portMap.values())
            {
                // Check if this sink is alive
                if (ipp.IsAlive())
                {
                    result &= sendData(bytes, size, ipp.ip, ipp.port);
                }
                else //Remove it.
                {
                    sinksToDelete.add(ipp.key);
                }
            }
            for (String key : sinksToDelete)
            {
                dict.portMap.remove(key);
            }
        }
        return result;
    }

    public synchronized void addSink(String topic, String ip, int port, boolean staticRoute) throws IOException
    {
        IpPortPair ipp = new IpPortPair(InetAddress.getByName(ip), port, staticRoute);

        if (this.topicSinkMap.containsKey(topic))
        {
            Entry_T dict = this.topicSinkMap.get(topic);

            //If created as static route, we only add sinks that are static
            if ((!dict.staticRoute) || (dict.staticRoute && staticRoute))
            {
                if (dict.portMap.containsKey(ipp.key))
                {
                    dict.portMap.get(ipp.key).FeedWatchdog(staticRoute);
                }
                else
                {
                    dict.portMap.put(ipp.key, ipp);
                }
            }
        }
        else
        {
            Entry_T dict = new Entry_T(staticRoute);
            dict.portMap.put(ipp.key, ipp);
            this.topicSinkMap.put(topic, dict);
        }
    }

    private class IpPortPair
    {
        public InetAddress ip;
        public int port;
        public String key;

        private boolean alwaysAlive = false;
        private long lastTimeAlive;
        private long ALIVE_TIMEOUT = 3000;  // [ms]

        public IpPortPair(InetAddress ip, int port, boolean alwaysAlive)
        {
            this.ip = ip;
            this.port = port;
            this.key = ip.toString() + ":" + port;
            this.alwaysAlive = alwaysAlive;
            this.lastTimeAlive = System.currentTimeMillis();
        }

        public boolean IsAlive()
        {
            return this.alwaysAlive || ((System.currentTimeMillis() - this.lastTimeAlive) < ALIVE_TIMEOUT);
        }

        public void FeedWatchdog(boolean alwaysAlive)
        {
            this.alwaysAlive |= alwaysAlive;
            this.lastTimeAlive = System.currentTimeMillis();
        }
    }

}
